File: python/cosmiclens_validate/sis.py
from __future__ import annotations


def sis_images(beta: float, theta_e: float) -> list[float]:
    if abs(beta) >= theta_e:
        sign = 1.0 if beta >= 0 else -1.0
        return [beta + sign * theta_e]
    return [beta + theta_e, beta - theta_e]


if __name__ == "__main__":
    print({"inside": sis_images(0.3, 1.0), "outside": sis_images(1.3, 1.0)})

File: packages/physics-core/src/models/sis.ts
import { EPS } from '../constants';
import type { Vec2 } from '../types';
import { norm, sub } from '../vector';

export function sisDeflection(theta: Vec2, center: Vec2, thetaE: number): Vec2 {
  const d = sub(theta, center);
  const r = Math.max(norm(d), EPS);
  return [thetaE * d[0] / r, thetaE * d[1] / r];
}

export function sisPotential(theta: Vec2, center: Vec2, thetaE: number): number {
  const d = sub(theta, center);
  return thetaE * Math.max(norm(d), EPS);
}

export function sisImagePositions1D(beta: number, thetaE: number): number[] {
  if (Math.abs(beta) >= thetaE) {
    return [beta + Math.sign(beta || 1) * thetaE];
  }
  return [beta + thetaE, beta - thetaE];
}
